File: practice-java/fundamentals/src/Herencia.java
public class Herencia {
    public static void main(String[] args) {

        // ===== CREAR INSTANCIAS DE CLASES CON HERENCIA =====

        System.out.println("===== HERENCIA Y POLIMORFISMO =====");

        Animal animal = new Animal("Genérico");
        animal.hacerSonido();
        animal.descripcion();

        System.out.println();

        // crear instancia de Perro (hereda de Animal)
        Perro perro = new Perro("Rocky", "Labrador");
        perro.hacerSonido();
        perro.descripcion();
        perro.traerObjeto();

        System.out.println();

        // crear instancia de Gato (hereda de Animal)
        Gato gato = new Gato("Miau", "Persa");
        gato.hacerSonido();
        gato.descripcion();
        gato.rasgunar();

        System.out.println();

        // crear instancia de Pajaro (hereda de Animal)
        Pajaro pajaro = new Pajaro("Tweety", "Canario");
        pajaro.hacerSonido();
        pajaro.descripcion();
        pajaro.volar();

        // ===== POLIMORFISMO: TRATAR OBJETOS COMO SU CLASE PADRE =====

        System.out.println("\n===== POLIMORFISMO =====");

        Animal[] animales = {
            new Perro("Bobby", "Pastor Alemán"),
            new Gato("Whiskers", "Siamés"),
            new Pajaro("Lolo", "Loro"),
            new Animal("Animal genérico")
        };

        System.out.println("Polimorfismo: llamar hacerSonido() en cada animal:");
        for (Animal a : animales) {
            a.hacerSonido();
        }

    }
}

// ===== CLASE PADRE (SUPERCLASE) =====

class Animal {
    // atributos
    protected String nombre; // protected permite acceso en subclases

    // constructor
    public Animal(String nombre) {
        this.nombre = nombre;
    }

    // método que puede ser sobrescrito (@Override)
    public void hacerSonido() {
        System.out.println(this.nombre + " hace un sonido genérico");
    }

    // método que muestra información
    public void descripcion() {
        System.out.println("Soy un animal llamado " + this.nombre);
    }
}

// ===== CLASE HIJA: PERRO =====

class Perro extends Animal {
    private String raza;

    // constructor del Perro
    public Perro(String nombre, String raza) {
        super(nombre); // llamar al constructor de la clase padre
        this.raza = raza;
    }

    // @Override: indica que este método reemplaza el método de la clase padre
    @Override
    public void hacerSonido() {
        System.out.println(this.nombre + " ladra: ¡Guau guau!");
    }

    // método específico de Perro
    public void traerObjeto() {
        System.out.println(this.nombre + " va a traer el objeto");
    }

    // sobrescribir el método descripcion
    @Override
    public void descripcion() {
        System.out.println("Soy un perro de raza " + this.raza + " llamado " + this.nombre);
    }
}

// ===== CLASE HIJA: GATO =====

class Gato extends Animal {
    private String raza;

    // constructor del Gato
    public Gato(String nombre, String raza) {
        super(nombre);
        this.raza = raza;
    }

    // @Override: sobrescribir el método hacerSonido
    @Override
    public void hacerSonido() {
        System.out.println(this.nombre + " maúlla: ¡Miau miau!");
    }

    // método específico de Gato
    public void rasgunar() {
        System.out.println(this.nombre + " está raspando con sus garras");
    }

    // sobrescribir el método descripcion
    @Override
    public void descripcion() {
        System.out.println("Soy un gato de raza " + this.raza + " llamado " + this.nombre);
    }
}

// ===== CLASE HIJA: PAJARO =====

class Pajaro extends Animal {
    private String especie;

    // constructor del Pajaro
    public Pajaro(String nombre, String especie) {
        super(nombre);
        this.especie = especie;
    }

    // @Override: sobrescribir el método hacerSonido
    @Override
    public void hacerSonido() {
        System.out.println(this.nombre + " canta: ¡Pío pío!");
    }

    // método específico de Pajaro
    public void volar() {
        System.out.println(this.nombre + " está volando por el cielo");
    }

    // sobrescribir el método descripcion
    @Override
    public void descripcion() {
        System.out.println("Soy un " + this.especie + " llamado " + this.nombre);
    }
}

File: practice-cpp/fundamentals/src/herencia.cpp
#include <iostream>

using namespace std;

// ===== CLASE BASE: ANIMAL =====

class Animal {
protected:
    string nombre;

public:
    Animal(string nombre) {
        this->nombre = nombre;
    }

    virtual void hacerSonido() {
        cout << nombre << " hace un sonido genérico\n";
    }

    virtual void descripcion() {
        cout << "Soy un animal llamado " << nombre << "\n";
    }

    virtual ~Animal() {
        // Destructor virtual (buena práctica en clases base)
    }
};

// ===== CLASE DERIVADA: PERRO =====

class Perro : public Animal {
private:
    string raza;

public:
    Perro(string nombre, string raza) : Animal(nombre) {
        this->raza = raza;
    }

    void hacerSonido() override {
        cout << nombre << " ladra: ¡Guau guau!\n";
    }

    void traerObjeto() {
        cout << nombre << " va a traer el objeto\n";
    }

    void descripcion() override {
        cout << "Soy un perro de raza " << raza << " llamado " << nombre << "\n";
    }
};

// ===== CLASE DERIVADA: GATO =====

class Gato : public Animal {
private:
    string raza;

public:
    Gato(string nombre, string raza) : Animal(nombre) {
        this->raza = raza;
    }

    void hacerSonido() override {
        cout << nombre << " maúlla: ¡Miau miau!\n";
    }

    void rasgunar() {
        cout << nombre << " está raspando con sus garras\n";
    }

    void descripcion() override {
        cout << "Soy un gato de raza " << raza << " llamado " << nombre << "\n";
    }
};

// ===== CLASE DERIVADA: PAJARO =====

class Pajaro : public Animal {
private:
    string especie;

public:
    Pajaro(string nombre, string especie) : Animal(nombre) {
        this->especie = especie;
    }

    void hacerSonido() override {
        cout << nombre << " canta: ¡Pío pío!\n";
    }

    void volar() {
        cout << nombre << " está volando por el cielo\n";
    }

    void descripcion() override {
        cout << "Soy un " << especie << " llamado " << nombre << "\n";
    }
};

// ===== MAIN =====

int main() {

    // ===== CREAR INSTANCIAS DE CLASES CON HERENCIA =====

    cout << "\n===== HERENCIA Y POLIMORFISMO =====\n";

    Animal animal("Genérico");
    animal.hacerSonido();
    animal.descripcion();

    cout << "\n";

    Perro perro("Rocky", "Labrador");
    perro.hacerSonido();
    perro.descripcion();
    perro.traerObjeto();

    cout << "\n";

    Gato gato("Miau", "Persa");
    gato.hacerSonido();
    gato.descripcion();
    gato.rasgunar();

    cout << "\n";

    Pajaro pajaro("Tweety", "Canario");
    pajaro.hacerSonido();
    pajaro.descripcion();
    pajaro.volar();

    // ===== POLIMORFISMO: TRATAR OBJETOS COMO SU CLASE BASE =====

    cout << "\n===== POLIMORFISMO =====\n";

    // Array de punteros a Animal
    Animal* animales[] = {
        &perro,
        &gato,
        &pajaro,
        &animal
    };

    cout << "Polimorfismo: llamar hacerSonido() en cada animal:\n";
    for (int i = 0; i < 4; i++) {
        animales[i]->hacerSonido();
    }

    cout << "\n";

    cout << "Polimorfismo: llamar descripcion() en cada animal:\n";
    for (int i = 0; i < 4; i++) {
        animales[i]->descripcion();
    }

    // ===== CAST A TIPO DERIVADO (DOWNCAST) =====

    cout << "\n===== DOWNCAST: CONVERTIR PUNTERO BASE A DERIVADA =====\n";

    Animal* ptrAnimal = &perro;

    // Downcast: convertir puntero base a puntero derivado
    // En C++ usaríamos dynamic_cast en entornos reales
    Perro* ptrPerro = static_cast<Perro*>(ptrAnimal);

    cout << "Acceso a método específico de Perro:\n";
    ptrPerro->traerObjeto();

    return 0;
}
